rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/explorer/AssetsRoot.kt

```diff
@@ -0,0 +1,137 @@
+            RdSolutionState.ReadyWithWarnings -> presentProjectsCount(presentation, descriptor.projectsCount, true)
+            else -> {}
+        }
+    }
+
+    private fun getAggregateSolutionState(descriptor: RdSolutionDescriptor): RdSolutionState {
+        var state = descriptor.state
+
+        // Solution loading/synchronizing takes precedence
+        if (state == RdSolutionState.Loading || state == RdSolutionState.Sync) {
+            return state
+        }
+
+        state = RdSolutionState.Ready
+        val children = solutionNode.getChildren(false, false)
+        for (child in children) {
+            if (child.isProject()) {
+                val projectDescriptor = child.descriptor as? RdProjectDescriptor ?: continue
+
+                // Aggregate project loading and sync. Loading takes precedence over sync
+                if (projectDescriptor.state == RdProjectState.Loading) {
+                    state = RdSolutionState.Loading
+                }
+                else if (projectDescriptor.state == RdProjectState.Sync && state != RdSolutionState.Loading) {
+                    state = RdSolutionState.Sync
+                }
+            }
+        }
+
+        // Make sure we don't miss solution ReadWithErrors
+        if (state == RdSolutionState.Ready) {
+            state = descriptor.state
+        }
+
+        return state
+    }
+
+    private fun presentProjectsCount(presentation: PresentationData, count: RdProjectsCount, showZero: Boolean) {
+        if (count.total == 0 && !showZero) return
+
+        var text = "${count.total} ${StringUtil.pluralize("project", count.total)}"
+        val unloadedCount = count.failed + count.unloaded
+        if (unloadedCount > 0) {
+            text += ", $unloadedCount unloaded"
+        }
+        presentation.addAdditionalText(text)
+    }
+
+    override fun isAlwaysExpand() = true
+
+    override fun calculateChildren(): MutableList<AbstractTreeNode<*>> {
+        val result = super.calculateChildren()
+        result.add(0, referenceRoot)
+        return result
+    }
+}
+
+class ReferenceRoot(project: Project) : AbstractTreeNode<Any>(project, key) {
+
+    companion object {
+        val key = Any()
+    }
+
+    override fun update(presentation: PresentationData) {
+        presentation.presentableText = "References"
+        presentation.setIcon(UnityIcons.Explorer.ReferencesRoot)
+    }
+
+    override fun getChildren(): MutableCollection<AbstractTreeNode<*>> {
+        val referenceNames = hashMapOf<String, ArrayList<ProjectModelNodeKey>>()
+        val visitor = object : ProjectModelNodeVisitor() {
+            override fun visitReference(node: ProjectModelNode): Result {
+                if (node.isAssemblyReference()) {
+                    val keys = referenceNames.getOrCreate(node.name) { _ -> arrayListOf() }
+                    keys.add(node.key)
+                }
+                return Result.Stop
+            }
+        }
+        visitor.visit(project!!)
```

Comment:
Could this ever be called when project is null?

Reply:
No, we only ever create the Unity explorer and nodes when we have a project.